File: helpdeskspringboot/src/main/java/com/ez/service/UserService.java
package com.ez.service;

import com.ez.payload.ChangePassword;
import com.ez.payload.EditProfile;
import com.ez.exception.*;
import com.ez.entity.User;

import javax.mail.MessagingException;
import javax.persistence.EntityNotFoundException;
import java.util.List;

public interface UserService {

    // find user by user id
    public User findById(Long id) throws EntityNotFoundException;

    // search users by pageNumber and based on the search criteria.
    // parameters:
    //  - pageNumber: page number
    //  - pageSize: page size
    //  - searchTerm: ID, email, firstName, lastName, phone
    //  - role: '', 'ROLE_CUSTOMER', 'ROLE_SUPPORTER', 'ROLE_ADMIN'
    //  - status: '', 'Active', 'Inactive'
    public List<User> searchUsers(int pageNumber, int pageSize, String searchTerm, String role, String status);

    // find user by email
    public User findUserByEmail(String email);

    // check whether user is inactive or not?
    // return:
    //  - not null: inactive user
    //  - null: not inactive user
    public User isInactiveUser(String email) throws BadDataException;

    // calculate total of users based on the search criteria.
    // based on this total of users we can calculate total pages
    public long getTotalOfUsers(String searchTerm, String role, String status);

    // create new user
    public User createUser(User user) throws MessagingException, BadDataException;

    // update existing user
    public User updateUser(User user) throws MessagingException, BadDataException, EntityNotFoundException;

    // update user profile
    public User updateProfile(EditProfile editProfile) throws MessagingException, EntityNotFoundException;

    // reset password in case user forgot his/her password
    public void resetPassword(String email) throws MessagingException, EntityNotFoundException;

    // change password
    public void changePassword(ChangePassword changePassword) throws MessagingException, EntityNotFoundException, BadDataException, BadDataException;

}

File: helpdeskangular/src/app/service/user.service.ts
import { HttpClient } from '@angular/common/http';
import { Injectable } from '@angular/core';
import { Observable } from 'rxjs';
import { environment } from 'src/environments/environment';
import { ChangePassword } from '../entity/ChangePassword';
import { CustomHttpRespone } from '../entity/CustomHttpRespone';
import { EditProfile } from '../entity/EditProfile';
import { ResetPassword } from '../entity/ResetPassword';
import { User } from '../entity/User';

@Injectable({
  providedIn: 'root'
})
export class UserService {


  // 'http://localhost:8080'
  host = environment.apiUrl;

  // number of users per page(default = 5)
  pageSize = environment.pageSize;

  constructor(private http: HttpClient) { }

  // get users by pageNumber and based on the search criteria
  searchUsers(pageNumber: number, searchTerm: string, role: string, status: string): Observable<User[]> {

    return this.http.get<User[]>(
      `${this.host}/user-search?pageNumber=${pageNumber}&pageSize=${this.pageSize}&searchTerm=${searchTerm}&role=${role}&status=${status}`
    )
  }

  // calculate total of users for count total pages
  getTotalOfUsers(searchTerm: string, role: string, status: string): Observable<number> {

    // ex: http://localhost:8080/total-of-users?searchTerm=""&role=""&status=""
    return this.http.get<number>(
      `${this.host}/total-of-users?searchTerm=${searchTerm}&role=${role}&status=${status}`
    );

  }

  // create new user
  public createUser(user: User): Observable<User> {
    return this.http.post<User>(`${this.host}/user-create`, user);
  }

  // edit existing user
  public editUser(user: User): Observable<User> {
    return this.http.put<User>(`${this.host}/user-edit`, user);
  }

  // reset password
  public resetPassword(resetPassword: ResetPassword): Observable<CustomHttpRespone> {

    return this.http.put<CustomHttpRespone>(`${this.host}/reset-password`, resetPassword);

  }

  // update profile
  public updateProfile(editProfile: EditProfile): Observable<User> {
    return this.http.put<User>(`${this.host}/edit-profile`, editProfile);
  }

  // change password
  public changePassword(changePassword: ChangePassword): Observable<CustomHttpRespone> {
    return this.http.put<CustomHttpRespone>(`${this.host}/change-password`, changePassword);
  }

  // find user by user id
  findById(id: number): Observable<User> {
    return this.http.get<User>(`${this.host}/user-list/${id}`);
  }

}
